scheduler/reconcile.go

```diff
@@ -720,13 +725,31 @@ func (a *allocReconciler) computePlacements(group *structs.TaskGroup,
 
 			downgradeNonCanary: canaryState && !alloc.DeploymentStatus.IsCanary(),
 			minJobVersion:      alloc.Job.Version,
+			lost:               false,
 		})
 	}
 
-	// Hot path the nothing to do case
+	// Add replacements for lost allocs up to group.Count
 	existing := len(untainted) + len(migrate) + len(reschedule)
-	if existing >= group.Count {
-		return place
+
+	for _, alloc := range lost {
+		if existing >= group.Count {
+			// Reached desired count, do not replace remaining lost
+			// allocs
+			break
+		}
+
+		existing++
+		place = append(place, allocPlaceResult{
+			name:               alloc.Name,
+			taskGroup:          group,
+			previousAlloc:      alloc,
+			reschedule:         false,
+			canary:             alloc.DeploymentStatus.IsCanary(),
+			downgradeNonCanary: canaryState && !alloc.DeploymentStatus.IsCanary(),
+			minJobVersion:      alloc.Job.Version,
+			lost:               true,
+		})
```

Comment:
This is a pretty big change. Before lost allocs would be replaced in the block below that simply ensures `existing == group.Count`. Now `lost` allocs are treated like `rescheduled` allocs and replaced 1:1 so we can propagate task handles from `previous->new` later on.

Reply:
I was a little worried about what happens to lost allocs after their node was GC'd (for the case where the handle can't get placed immediately because of constraints, but then the original node gets GC'd), but it looks like `filterByTainted` already covers that case! 👍